include/rnp.h

```diff
@@ -48,85 +49,101 @@ __BEGIN_DECLS
+    char *    defkey;    /* default key id */
+    int       pswdtries; /* number of password tries, -1 for unlimited */
+
+    enum key_store_format_t key_store_format; /* keyring format */
+    union {
+        rnp_keygen_desc_t generate_key_ctx;
+    } action;
+} rnp_t;
+
+/* rnp initialization parameters : keyring pathes, flags, whatever else */
+typedef struct rnp_params_t {
+    unsigned enable_coredumps; /* enable coredumps: if it is allowed then they are disabled by
+                                  default to not leak confidential information */
+
+    int         passfd; /* password file descriptor */
+    const char *outs;   /* output stream : may be <stderr> , most likel these are subject for
+                           refactoring  */
+    const char *errs;   /* error stream : may be <stdout> */
+    const char *ress;   /* results stream : maye be <stdout>, <stderr> or file name/path */
+
+    enum key_store_format_t ks_format; /* format of the key store */
+    char *                  pubpath;   /* public keystore path */
+    char *                  secpath;   /* secret keystore path */
+    char *                  defkey;    /* default/preferred key id */
+} rnp_params_t;
+
 /* rnp operation context : contains additional data about the currently ongoing operation */
 typedef struct rnp_ctx_t {
+    rnp_t *        rnp;       /* rnp structure */
     char *         filename;  /* name of the input file to store in literal data packet */
     int64_t        filemtime; /* file modification time to store in literal data packet */
-    int            halg;      /* hash algorithm */
+    int64_t        sigcreate; /* signature creation time */
+    uint64_t       sigexpire; /* signature expiration time */
+    pgp_hash_alg_t halg;      /* hash algorithm */
     pgp_symm_alg_t ealg;      /* encryption algorithm */
     int            zalg;      /* compression algorithm used */
     int            zlevel;    /* compression level */
     int            overwrite; /* allow to overwrite output file if exists */
     int            armour;    /* use ASCII armour on output */
 } rnp_ctx_t;
 
-/* structure used to hold (key,value) pair information */
-typedef struct rnp_t {
-    unsigned  c;       /* # of elements used */
-    unsigned  size;    /* size of array */
-    char **   name;    /* key names */
-    char **   value;   /* value information */
-    void *    pubring; /* public key ring */
-    void *    secring; /* s3kr1t key ring */
-    void *    io;      /* the io struct for results/errs */
-    void *    passfp;  /* file pointer for password input */
-    rnp_ctx_t ctx;     /* current operation context */
-
-    enum key_store_format_t key_store_format; /* keyring format */
-    union {
-        rnp_keygen_desc_t generate_key_ctx;
-    } action;
-} rnp_t;
-
-/* begin and end */
-int rnp_init(rnp_t *);
+/* initialize rnp using the init structure  */
+int rnp_init(rnp_t *, rnp_params_t *);
```

Comment:
1. It has the same return type as before, the error code, as this is done for most function. Imho it is much better to change return types all at once then to have some functions returning int, and some - returning pgp_error_t, with different meaning for 0 and 1.

2. See 1.
3. Yeah, it should, thanks.